HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,91 @@
+# I throw in the towel.  This is a mess and I am not close to finishing.  
+
+import findspark
+findspark.init()
+import pyspark
+
+def BFS(graph, names, iterations=10, inspect=[], sc):
+    
+    # Rdd of path that will be added to.  
+    rdd = sc.parallelize([(inspect[0],0)], 2).cache()
+    
+    # This RDD is used intermediately to build a collection of connections to build on
+    namesRdd = rdd
+
+    # Iterator used in the while loop
+    i = 0
+       
+    acc = sc.accumulator(0)    
+
+    while acc.value == 0 and i <= iterations:
+
+        i+=1
+
+        # Use join to look up the connections of the next round of connections
+        namesRdd = graph.join(namesRdd)
+
+        # Use mapValues to take just the list of connections and remove the distance
+        namesRdd = namesRdd.map(lambda x: (x[0], x[1][0])).partitionBy(2)
+
+        # Take the connections and flatMap them into one big pool to consider
+        namesRdd = namesRdd.values().flatMap(lambda x: x)
+
+        # Reduce the connections to distinct entries
+        namesRdd = namesRdd.distinct()
+        
+        # Subtract out the names that are already in the rdd
+        namesRdd = namesRdd.subtract(rdd.keys())
+        
+        if namesRdd.isEmpty(): acc = sc.accumulator(1)   # If nothing was added, change the accumulator value to exit the loop
+        else:
+            # create tuples of the connection name and the current distance of the iteration
+            namesRdd = namesRdd.map(lambda x: (x, i)).partitionBy(2).cache()
+
+            # Union the connection tuples to the existing list, and repartition
+            rdd = rdd.union(namesRdd).partitionBy(2).cache()
+ 
```

Comment:
Instead of using iterations, you should have checked if you reached the end target (the target node was found).
This will also reduce the number of iterations that are performed.